File: src/VoiceBuddy.App/Services/DebugLog.cs
namespace VoiceBuddy.Services;

public enum LogDirection { Out, In }

public sealed record LogEntry(DateTime Timestamp, LogDirection Direction, string Kind, string Payload);

/// <summary>
/// In-memory bounded log of API traffic (REST + WebSocket). Zero cost when Enabled is false.
/// Raw audio chunks are gated behind <see cref="LogAudioChunks"/> because they arrive ~8×/s
/// and drown out everything else.
/// </summary>
public sealed class DebugLog
{
    private const int MaxEntries = 1000;

    private readonly List<LogEntry> _entries = new(MaxEntries + 64);
    private readonly object _lock = new();

    public bool Enabled { get; set; }
    public bool LogAudioChunks { get; set; }

    public event EventHandler<LogEntry>? EntryAdded;
    public event EventHandler? Cleared;

    public void Log(LogDirection dir, string kind, string payload)
    {
        if (!Enabled) return;
        var entry = new LogEntry(DateTime.Now, dir, kind, payload);
        lock (_lock)
        {
            _entries.Add(entry);
            if (_entries.Count > MaxEntries) _entries.RemoveRange(0, _entries.Count - MaxEntries);
        }
        EntryAdded?.Invoke(this, entry);
    }

    public void LogAudioChunk(int byteCount)
    {
        if (!Enabled || !LogAudioChunks) return;
        Log(LogDirection.Out, "source_media_chunk", $"{byteCount} bytes");
    }

    public IReadOnlyList<LogEntry> Snapshot()
    {
        lock (_lock) return _entries.ToArray();
    }

    public void Clear()
    {
        lock (_lock) _entries.Clear();
        Cleared?.Invoke(this, EventArgs.Empty);
    }
}

File: src/mac/VoiceBuddy/VoiceBuddy/Services/DebugLog.swift
import Foundation
import Combine

enum LogDirection: String {
    case out = "→"
    case `in` = "←"
}

struct LogEntry: Identifiable {
    let id = UUID()
    let timestamp: Date
    let direction: LogDirection
    let kind: String
    let payload: String
}

@MainActor
final class DebugLog: ObservableObject {
    @Published var enabled: Bool = false
    @Published var logAudioChunks: Bool = false
    @Published private(set) var entries: [LogEntry] = []
    private let cap = 1000

    func log(_ dir: LogDirection, kind: String, payload: String) {
        guard enabled else { return }
        let entry = LogEntry(timestamp: Date(), direction: dir, kind: kind, payload: payload)
        entries.append(entry)
        if entries.count > cap {
            entries.removeFirst(entries.count - cap)
        }
    }

    func audioChunk(bytes: Int) {
        guard enabled, logAudioChunks else { return }
        log(.out, kind: "source_media_chunk", payload: "\(bytes) bytes")
    }

    func clear() {
        entries.removeAll()
    }
}
